src/table_properties.rs

```diff
@@ -0,0 +1,244 @@
+
+        let factory: &mut F = &mut *(raw_self.cast());
+        let collector = Box::new(factory.create(context));
+
+        ffi::rocksdb_table_properties_collector_create(
+            Box::into_raw(collector).cast(),
+            Some(TablePropertiesCollectorCallback::<F::Collector>::destructor),
+            Some(TablePropertiesCollectorCallback::<F::Collector>::add_user_key),
+            Some(TablePropertiesCollectorCallback::<F::Collector>::block_add),
+            Some(TablePropertiesCollectorCallback::<F::Collector>::finish),
+            Some(TablePropertiesCollectorCallback::<F::Collector>::get_readable_properties),
+            Some(TablePropertiesCollectorCallback::<F::Collector>::name),
+        )
+    }
+
+    unsafe extern "C" fn name(raw_self: *mut c_void) -> *const c_char {
+        let factory = &*(raw_self.cast_const() as *const F);
+        factory.name().as_ptr()
+    }
+
+    unsafe extern "C" fn destructor(raw_self: *mut c_void) {
+        drop(Box::from_raw(raw_self as *mut F));
+    }
+}
+
+struct TablePropertiesCollectorCallback<C>
+where
+    C: TablePropertiesCollector,
+{
+    _marker: PhantomData<C>,
+}
+
+impl<C> TablePropertiesCollectorCallback<C>
+where
+    C: TablePropertiesCollector,
+{
+    unsafe extern "C" fn destructor(raw_collector: *mut c_void) {
+        drop(Box::from_raw(raw_collector as *mut C));
+    }
+
+    unsafe extern "C" fn name(raw_collector: *mut c_void) -> *const c_char {
+        let collector: &mut C = &mut *(raw_collector.cast());
+        collector.name().as_ptr()
+    }
+
+    unsafe extern "C" fn add_user_key(
+        raw_collector: *mut c_void,
+        key_ptr: *const c_char,
+        key_len: size_t,
+        value_ptr: *const c_char,
+        value_len: size_t,
+        entry_type: c_int,
+        seq: u64,
+        file_size: u64,
+    ) -> bool {
+        let collector: &mut C = &mut *(raw_collector.cast());
+
+        let key = slice::from_raw_parts(key_ptr as *const u8, key_len);
+        let value = slice::from_raw_parts(value_ptr as *const u8, value_len);
+        let entry_type = mem::transmute::<c_int, EntryType>(entry_type);
+
+        collector.add_user_key(key, value, entry_type, seq, file_size)
+    }
+
+    unsafe extern "C" fn block_add(
+        raw_collector: *mut c_void,
+        block_uncompressed_bytes: u64,
+        block_compressed_bytes_fast: u64,
+        block_compressed_bytes_slow: u64,
+    ) {
+        let collector: &mut C = &mut *(raw_collector.cast());
+        collector.block_add(
+            block_uncompressed_bytes,
+            block_compressed_bytes_fast,
+            block_compressed_bytes_slow,
+        );
+    }
+
+    unsafe extern "C" fn finish(
+        raw_collector: *mut c_void,
+        user_collected_properties: *mut ffi::rocksdb_user_collected_properties_t,
+    ) -> bool {
+        let collector: &mut C = &mut *(raw_collector.cast());
+
+        let mut props = HashMap::new();
+        collector.finish(&mut props);
```

Comment:
Same here. If `finish` returned an iterator, then we wouldn't have to allocate a `HashMap`.

Reply:
Theoretically, the Rust `TablePropertiesCollector` wouldn't have to return a `CString` but the conversion could happen here in the callback.